### ISSUE TITLE
Compilation crashes when nodes under heavy load

### ISSUE THREAD

**Issue by sdolezal** [icecc/icecream#336]
When compiling with icecream we are experiencing the following exception:
`ICECC[32515] 10:05:54: replied not with use_cs 0`
`ICECC[32515] 10:05:54: got exception 1 (this should be an exception!)`

or

`ICECC[5714] 12:57:02: reply was not expected use_cs 0`
`ICECC[5714] 12:57:02: got exception Error 1 - expected use_cs reply, but got 0 instead (this should be an exception!)`

and also the aforementioned exception 15  (#228)

We're using icecream 1.1 for compiling in a cluster of machines running openSuse 42.1 and 42.3. This particular problem happens almost everytime when compiling (and linking) one of our bigger projects, which needs quite a long time for linking the main binary and our test-binaries. Automated builds are started simultaneously on a 42.1 and 42.3 System using icecream. It seems, that the exceptions happen preferrably during linking times on the local nodes which started the build.

Since nobody mentioned the "use_cs" errors before, hopefully this will provide a different perspective on the problem.

**Comment by HenryMiller1** [comment]
I have seen this often myself. I don't have any idea how to debug
it though.
We never see it on our ci systems of powerful machines (16+ cores , 128
gig ram). We see it all the time on a different network of more normal
desktops (6 cores, 12 gig, running in a vm). I don't know what about the
above difference is important though.

--
  Henry Miller
  hank@millerfarm.com



On Fri, Jan 12, 2018, at 4:42 AM, Stephan Dolezal wrote:
> When compiling with icecream we are experiencing the following
> exception:> ICECC[32515] 10:05:54: replied not with use_cs 0
> ICECC[32515] 10:05:54: got exception 1 (this should be an exception!)> or


> ICECC[5714] 12:57:02: reply was not expected use_cs 0
> ICECC[5714] 12:57:02: got exception Error 1 - expected use_cs reply,
> but got 0 instead (this should be an exception!)> and also the aforementioned exception 15  #(228)


> We're using icecream 1.1 for compiling in a cluster of machines
> running openSuse 42.1 and 42.3. This particular problem happens
> almost everytime when compiling (and linking) one of our bigger
> projects, which needs quite a long time for linking the main binary
> and our test-binaries. Automated builds are started simultaneously on
> a 42.1 and 42.3 System using icecream. It seems, that the exceptions
> happen preferrably during linking times on the local nodes which
> started the build.> Since nobody mentioned the "use_cs" errors before, hopefully this will
> provide a different perspective on the problem.> — You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub[1], or mute the
> thread[2].> 

Links:

  1. https://github.com/icecc/icecream/issues/336

**Comment by llunak** [comment]
Could you try if you can reproduce the problem also using the git version?
Can you be more specific about "quite a long time for linking" ?

**Comment by sdolezal** [comment]
We are building our project as static lib and then link our tests against it. Linking of the main executable takes about 10-20s, linking for each test (~100) takes about 1-4s

During the last week we fiddled with the icecream configs on our machines. I suspected that the machines, which start a build are causing the problem when simultaneously used as remote client for other builds. So we changed the setting ICECREAM_ALLOW_REMOTE to "false" on those machines which start builds. From what I can tell until now, this seems to have improved the situation and we will keep testing this config during the week.

I will make a suggestion to build the current icecream version from github and test it in-house.

**Comment by sdolezal** [comment]
Our config changes did not solve the problem. We will test the latest icecream version from github during the next week.

**Comment by bookkey** [comment]
We now checked the latest version from github (i'm a colleague of @sdolezal).
I'm sorry, but this version has another issue.
We have the following behaviour, but we didn't find out, what's the difference (yet).
If we compile a C++ project without Qt, it works.

If we try to compile any project with Qt (5.7.0) we got the following debug output (ICECC_DEBUG="debug")
`
ICECC[5031] 09:52:16: command needs stdout/stderr workaround, recompiling locally
ICECC[5031] 09:52:16: timeout <= 0
ICECC[5031] 09:52:16: saw eof without complete msg! 1
ICECC[5031] 09:52:16: local build forced by remote exception: Error 102 - command needs stdout/stderr workaround, recompiling locally
ICECC[5031] 09:52:16: invoking: /usr/bin/g++-4.8 -DDSO_MAIN -DMAIN_LIB_VERSION="7.2.42" -DQT_CORE_LIB -DQT_DBUS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_PRINTSUPPORT_LIB -DQT_SCRIPT_LIB -DQT_USE_FAST_CONCATENATION -DQT_USE_FAST_OPERATOR_PLUS -DQT_WIDGETS_LIB -DQT_XML_LIB -DWITH_DBUS="1" -Dicqt_main_EXPORTS -I <removed> -isystem <removed> -Wall -fvisibility=hidden -std=c++11 -g -fPIC -fPIC -std=gnu++11 -c -fdirectives-only <cppFileRemoved> -o <OFileRemoved>
ICECC[5033] 09:52:16: </wait for cs: 825ms>
`
(I removed some pathes. See <removed>)
Do you have any idea?

**Comment by llunak** [comment]
The stdout/stderr workaround is that newer gcc versions also prints parts of the source when printing warnings/errors, but gcc needs the original source file for that to work, which isn't the case with icecream's remote compilation (gcc 4.8 is not actually that new I think, but Icecream doesn't check the version). Since that would alter the compilation result (as far as the reported messages go), if remote compilation detects any output from gcc, the compilation is forced again locally in order to get the proper output.

You can fix your warnings, use Clang or disable the show caret workaround (see icecc --help).

**Comment by jolz** [comment]
(i'm a colleague of @sdolezal, too ;-) )

tl;dr: **qt-5.9.0 does not compile with icecream**, some older qt versions work. gcc-preprocessor seems to fail. 

Sorry, this may be too much text for the bug tracker. I do not know how to describe this shorter. 


@llunak Thanks a lot, this helps debugging. 
We included ``-fno-diagnostics-show-caret`` in out CMakeLists.txt as CMAKE_CXX_FLAGS. 

This works for our old qt4 projects, but qt5 continues to fail to compile with errors like these:
```
In file included from .../qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtCore/qvariant.h:44:0,
                 from.../qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtCore/qabstractitemmodel.h:43,
                 from .../qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtCore/QAbstractTableModel:1,
                 from <myfile.h>,
                 from <mymyhile.cpp>:
.../qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtCore/qbytearray.h:291:81: error: expected ';' at end of member declaration
.../qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtCore/qbytearray.h:291:87: error: 'Q_REQUIRED_RESULT' does not name a type
.../qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtCore/qbytearray.h:292:82: error: expected ';' at end of member declaration
.../qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtCore/qbytearray.h:292:88: error: 'Q_REQUIRED_RESULT' does not name a type

```

We do not see these warnings with ``ICECC=no``. 



This is my minimal test project: 
CMakeLists.txt
```
cmake_minimum_required(VERSION 2.8.11)
project(testproject)
set(CMAKE_INCLUDE_CURRENT_DIR ON)
set(CMAKE_AUTOMOC ON)
set(CMAKE_CXX_FLAGS "-Wextra -Wall  -fno-diagnostics-show-caret")
find_package(Qt5Widgets)
add_executable(helloworld WIN32 main.cpp)
target_link_libraries(helloworld Qt5::Widgets)
```
main.cpp
```
#include <QtCore/QStringList>
#include <QtCore/QDir>
#include <QtCore/QCoreApplication>


#include <QDebug>
#include <QApplication>
#include <QMainWindow>

int main(int argc, char *argv[])
{	
    QApplication app(argc, argv);

	qDebug() << "hello world from qt" << QT_VERSION_STR;
	QMainWindow w;
	w.show();
    return app.exec();
}
```

do.sh
``` 
#!/bin/bash
# this must be another hostname in your icecc-cluster:
export ICECC_PREFERRED_HOST=other_host
export PATH=/usr/lib/icecc/bin:$PATH 

export ICECC_DEBUG=debug 
rm -rf build/ 
mkdir build 
cd build 
cmake .. 
make VERBOSE=1

echo "_____________________________"
timeout 3s ./helloworld
```

Testproject (see below) - please compile with
```
do.sh # good (uses system-qt, i.e. qt-5.6.0 in OpenSuSE 42.3)
CMAKE_PREFIX_PATH=<qt-5.2.0-dir>/5.2.0/gcc_64 ./do.sh # good 
ICECC_CC=/usr/bin/gcc-7 ICECC_CXX=/usr/bin/g++-7 CMAKE_PREFIX_PATH=<qt-5.2.0-dir>/5.2.0/gcc_64 ./do.sh # good 
CMAKE_PREFIX_PATH=<qt-5.7.0-dir>/5.7/gcc_64 ./do.sh # FAIL
ICECC=no CMAKE_PREFIX_PATH=<qt-5.7.0-dir>/5.7/gcc_64 ./do.sh # good, but no icecream
```
The test is compiled (against qt-opensource from binary archives from https://download.qt.io/archive/qt/
gcc-version: 4.8.5, 5.3.1, 7.2.1
not all gcc-versions are printing the " Q_REQUIRED_RESULT "- message, some fail with ```got exception Error 14 - error reading message from remote```


with a patched icecc client (client/cpp.cpp) we got the preprocessor command: 
```forking /usr/bin/c++ -I/data/tmp/hltestcase_icecc/test_a/build -I/data/tmp/hltestcase_icecc/test_a -isystem /qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include -isystem /qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtWidgets -isystem /qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtGui -isystem /qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/include/QtCore -isystem /qt-linux-opensource-5.7.0-x86_64/5.7/gcc_64/./mkspecs/linux-g++ -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_WIDGETS_LIB -Wextra -Wall -fno-diagnostics-show-caret -fPIC -std=gnu++11 -E /data/tmp/hltestcase_icecc/test_a/main.cpp -fdirectives-only```

The resulting file (stdout) can has several ```#define Q_REQUIRED_RESULT``` and  ```#undef Q_REQUIRED_RESULT```, but the Q_REQUIRED_RESULT is used while undef'd.

So we assume, the gcc-preprocessor in stand-alone-mode seems to fail (?).

**Comment by jolz** [comment]
WORKAROUND: 
We just downgraded icecc back to 1.0.1 on our workstations (SuSE 42.3-default), compilation now works for all qt-versions. (some hosts, which do not start compilation, have still 1.1-HEAD)

**Comment by llunak** [comment]
Apparently gcc's -fdirectives-only mishandles the #pragma macro_push/pop around qbytearray.h line 284. The idea is that

g++ -E -fdirectives-only a.cpp | g++ -Wall -fpreprocessed -fdirectives-only -x c++ -

should be equivalent to

g++ a.cpp -c

but in this case it isn't. You can try both cases manually with your source or others.

So this is a gcc bug.

You can force the old mode that doesn't use -fdirectives-only by setting ICECC_REMOTE_CPP=0.

**Comment by jolz** [comment]
That's it - Thank You :-))

Summary for future reference: Our problem was a gcc bug processing with qt-5.7 headers.

To use icecc with gcc+qt5.7 we now use: 
- ICECC_CARET_WORKAROUND=0 or use gcc-flag "-fno-diagnostics-show-caret"
- ICECC_REMOTE_CPP=0
- have fun :-))

**Comment by sdolezal** [comment]
We've tested the changes (commit 8c75222)  for a week and it feels like the issue (see top entry) is happening less often but it is still there.

The machines in our cluster (about 40) have a job limit of 2. From time to time we still experience the following fault on the machines which started the build (they are also used as remote clients at the same time):
ICECC[8964] 15:18:10: reply was not expected use_cs 0
ICECC[8964] 15:18:10: got exception Error 1 - expected use_cs reply, but got 0 instead (this should be an exception!)

**Comment by sdolezal** [comment]
A few days ago we switched from static linking our automated tests with the main library to dynamic linking which reduced linking time noticeably. Since then we did not experience the aforementioned crashes anymore. My guess is, that the crashes were caused by the heavy load when static linking our tests and thus the compile client being unresponsive for a short time.

Since this solution works for us, I will close the ticket.

**Comment by jvicenti** [comment]
> That's it - Thank You :-))
> 
> Summary for future reference: Our problem was a gcc bug processing with qt-5.7 headers.
> 
> To use icecc with gcc+qt5.7 we now use:
> 
> * ICECC_CARET_WORKAROUND=0 or use gcc-flag "-fno-diagnostics-show-caret"
> * ICECC_REMOTE_CPP=0
> * have fun :-))

In qbytearray.h and qstring.h, you can replace the #pragma push_macro("Q_REQUIRED_RESULT") section with #define Q_REQUIRED_RESULT_NONE and make the appropriate code changes to use that #define. That should allow you to build with icecc without needing ICECC_CARET_WORKAROUND=0 and ICECC_REMOTE_CPP=0. This should improve build times significantly in certain use cases.
